File: 2414100208-黄楚民/第四周作业/app/src/main/java/com/example/camera/MainActivity.java
package com.example.camera; // 注意：如果你的包名不是这个，请修改这一行

import android.Manifest;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.camera.core.CameraSelector;
import androidx.camera.core.ImageAnalysis;
import androidx.camera.core.ImageCapture;
import androidx.camera.core.ImageCaptureException;
import androidx.camera.core.ImageProxy;
import androidx.camera.core.Preview;
import androidx.camera.lifecycle.ProcessCameraProvider;
import androidx.camera.view.PreviewView;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import com.google.common.util.concurrent.ListenableFuture;

import java.io.File;
import java.nio.ByteBuffer;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class MainActivity extends AppCompatActivity {

    private PreviewView viewFinder;
    private TextView tvLowLight;
    private ImageCapture imageCapture;
    private ExecutorService cameraExecutor;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        viewFinder = findViewById(R.id.viewFinder);
        tvLowLight = findViewById(R.id.tv_low_light);
        Button captureButton = findViewById(R.id.capture_button);
        cameraExecutor = Executors.newSingleThreadExecutor();

        // 检查权限并启动相机
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED) {
            startCamera();
        } else {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.CAMERA}, 10);
        }

        // 拍照按钮点击事件
        captureButton.setOnClickListener(v -> {
            File photoFile = new File(getExternalFilesDir(null), System.currentTimeMillis() + ".jpg");
            ImageCapture.OutputFileOptions options = new ImageCapture.OutputFileOptions.Builder(photoFile).build();
            imageCapture.takePicture(options, ContextCompat.getMainExecutor(this), new ImageCapture.OnImageSavedCallback() {
                @Override
                public void onImageSaved(@NonNull ImageCapture.OutputFileResults results) {
                    Toast.makeText(MainActivity.this, "保存成功: " + photoFile.getName(), Toast.LENGTH_SHORT).show();
                }
                @Override
                public void onError(@NonNull ImageCaptureException e) { e.printStackTrace(); }
            });
        });
    }

    private void startCamera() {
        ListenableFuture<ProcessCameraProvider> future = ProcessCameraProvider.getInstance(this);
        future.addListener(() -> {
            try {
                ProcessCameraProvider provider = future.get();

                // 预览
                Preview preview = new Preview.Builder().build();
                preview.setSurfaceProvider(viewFinder.getSurfaceProvider());

                // 拍照
                imageCapture = new ImageCapture.Builder().build();
                ImageAnalysis analysis = new ImageAnalysis.Builder()
                        .setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST).build();
                analysis.setAnalyzer(cameraExecutor, image -> {
                    double luma = getLuma(image);
                    runOnUiThread(() -> tvLowLight.setVisibility(luma < 50 ? View.VISIBLE : View.GONE));
                    image.close();
                });

                provider.unbindAll();
                provider.bindToLifecycle(this, CameraSelector.DEFAULT_BACK_CAMERA, preview, imageCapture, analysis);
            } catch (Exception e) { e.printStackTrace(); }
        }, ContextCompat.getMainExecutor(this));
    }

    private double getLuma(ImageProxy image) {
        ByteBuffer buffer = image.getPlanes()[0].getBuffer();
        byte[] data = new byte[buffer.remaining()];
        buffer.get(data);
        long sum = 0;
        for (byte b : data) sum += (b & 0xFF);
        return sum / (double) data.length;
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == 10 && grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            startCamera();
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        cameraExecutor.shutdown();
    }
}

File: 2414100138-张秀华/第十一周作业/2414100138-张秀华-第十一周代码/app/src/main/java/com/example/downloadpicture/view/MainActivity.kt
package com.example.downloadpicture.view

import android.graphics.ColorFilter
import android.graphics.PorterDuff
import android.os.Bundle
import android.widget.Button
import android.widget.ProgressBar
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import com.example.downloadpicture.R
import com.example.downloadpicture.viewmodel.DownloadViewModel

class MainActivity : AppCompatActivity() {
    private lateinit var vm: DownloadViewModel
    private lateinit var progressBar: ProgressBar
    private lateinit var tvText: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        progressBar = findViewById(R.id.progress_bar)
        tvText = findViewById(R.id.tv_text)
        progressBar.progressDrawable.colorFilter = android.graphics.PorterDuffColorFilter(
            android.graphics.Color.parseColor("#2196F3"),
            PorterDuff.Mode.SRC_IN
        )
        vm = ViewModelProvider(this)[DownloadViewModel::class.java]

        //下载按钮
        findViewById<Button>(R.id.btn_start).setOnClickListener {
            vm.startDownload(this@MainActivity)
        }

        //下载进度
        vm.progressLiveData.observe(this) { progress ->
            tvText.text = "当前下载进度：$progress %"
            progressBar.progress = progress
        }

        //错误提示
        vm.errorLiveData.observe(this){errMsg->
            tvText.append("\n$errMsg")
        }
    }
}